README.md
# Hashcash
Hashcash is a proof-of-work algorithm, which has been used as a denial-of-service counter measure technique in a 
number of systems.

A hashcash stamp constitutes a proof-of-work which takes a parameterizable amount of work to compute for the sender. 
The recipient (and indeed anyone as it is publicly auditable) can verify received hashcash stamps efficiently. 
Hashcash was invented by [Adam Back][cypherspace] in [1997][papers]

At this point it is most widely used as the [bitcoin][bitcoin] mining function.

The email anti-spam tool, like the proof-of-work algorithm, is also called hashcash and is used to create stamps to 
attach to mail to add a micro-cost to sending mail to deter spamming. The main use of the hashcash stamp is as a 
white-listing hint to help hashcash users avoid losing email due to content based and blacklist based anti-spam systems.

Hashcash source code includes a library form, and also the algorithm is extremely simple to code from scratch with the 
availability of a hash library. Verification can be done by a human eye (count leading 0s) even with availability of 
common preinstalled command line tools such as sha1sum. The algorithm works with a cryptographic hash, such as SHA1, 
SHA256 or coming SHA3 that exhibits 2nd-preimage resistance. Note that 2nd-preimage resistance is a stronger hash 
property than the collision resistance property.


## Usage

A simple example:

```dart
import 'package:hashcash/hashcash_dart.dart';

main() {
  var stamp = Hashcash.mint('dev@konstantinullrich.de');
  var checked = Hashcash.check(stamp, resource: 'dev@konstantinullrich.de');
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[cypherspace]: http://www.cypherspace.org/adam/
[papers]: http://hashcash.org/papers/
[bitcoin]: https://bitcoin.it/
[tracker]: https://github.com/konstantinullrich/hashcash/issues


CHANGELOG.md
## 1.1.0
- Nullsafety and Dart 3 release

## 1.0.1
- Finalize the nullsafety release

## 1.0.0-nullsafety
- Update dependencies

## 0.1.1
- Update dependencies
- Update tests

## 0.1.0
- adjust naming

## 0.0.1
- Implement the Hashcash protocol Version 1 


test/hashcash_test.dart
import 'dart:io';

import 'package:hashcash_dart/hashcash_dart.dart';
import 'package:test/test.dart';

void main() {
  group('Hashcash Testgroup', () {
    final stamp = Hashcash.mint('dev@konstantinullrich.de', stampSeconds: true);

    test('Wrong Resource', () {
      expect(Hashcash.check(stamp, resource: 'test'), false);
    });

    test('Valid Resource', () {
      expect(Hashcash.check(stamp, resource: 'dev@konstantinullrich.de'), true);
    });

    test('Correct bits', () {
      expect(Hashcash.check(stamp, bits: 20), true);
    });

    test('Smaller bits', () {
      expect(Hashcash.check(stamp, bits: 15), false);
    });

    test('Valid Expiration', () {
      expect(Hashcash.check(stamp, checkExpiration: Duration(hours: 1)), true);
    });

    sleep(Duration(seconds: 10));

    test('Bad Expiration', () {
      expect(
          Hashcash.check(stamp, checkExpiration: Duration(seconds: 1)), false);
    });

    test('More Valid Expiration', () {
      expect(
          Hashcash.check(stamp, checkExpiration: Duration(minutes: 1)), true);
    });
  });
}


lib/src/minter.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart' as crypto;

/// Checks if you are awesome. Spoiler: you are.
class Minter {
  final int _version = 1;
  int _tries = 0;
  final String _chars =
      'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/=';

  /// Version of the implemented hashcash protocol
  /// Warning: This version has nothing to do with the version fo the Package
  /// stated in the pubspec.yaml file
  int get version => _version;

  int get tries => _tries;

  /// Mint a new hashcash stamp for 'resource' with 'bits' of collision
  ///
  ///    20 bits of collision is the default.
  ///
  ///    'ext' lets you add your own extensions to a minted stamp.  Specify an
  ///    extension as a string of form 'name1=2,3;name2;name3=var1=2,2,val'
  ///    FWIW, urllib.urlencode(dct).replace('&',';') comes close to the
  ///    hashcash extension format.
  ///
  ///    'saltchars' specifies the length of the salt used; this version defaults
  ///    8 chars, rather than the C version's 16 chars.  This still provides about
  ///    17 million salts per resource, per timestamp, before birthday paradox
  ///    collisions occur.  Really paranoid users can use a larger salt though.
  ///
  ///    'stamp_seconds' lets you add the option time elements to the datestamp.
  ///    If you want more than just day, you get all the way down to seconds,
  ///    even though the spec also allows hours/minutes without seconds.
  String mint(String resource,
      {int bits = 20,
      var now,
      String extension = '',
      int saltchars = 8,
      bool stamp_seconds = false}) {
    var timestamp = '';

    var challenge = <String>[
      version.toString(),
      bits.toString(),
      timestamp,
      resource,
      extension,
      _salt(saltchars)
    ];

    var mint = _mint(challenge.join(':'), bits);
    challenge.add(mint);
    return challenge.join(':');
  }

  /// Return a random string of specified length
  String _salt(int length) {
    var random = Random();
    var result = '';
    for (var i = 0; i < length; i++) {
      result += _chars[random.nextInt(_chars.length)];
    }
    return result;
  }

  /// Answer a 'generalized hashcash' challenge'
  ///
  /// Hashcash requires stamps of form 'ver:bits:date:res:ext:rand:counter'
  /// This internal function accepts a generalized prefix 'challenge',
  /// and returns only a suffix that produces the requested SHA leading zeros.
  ///
  /// NOTE: Number of requested bits is rounded up to the nearest multiple of 4
  String _mint(String challenge, int bits) {
    var counter = 0;
    var hex_digets = (bits / 4).ceil();
    var zeros = '0' * hex_digets;
    while (true) {
      var digest = crypto.sha1
          .convert(utf8.encode(challenge + ':' + counter.toRadixString(16)))
          .toString();
      if (digest.startsWith(zeros)) {
        _tries = counter;
        return counter.toRadixString(16);
      }
      counter++;
    }
  }
}


lib/src/hashcash.dart
import 'dart:math';

import 'package:crypto/crypto.dart' as crypto;

class Hashcash {
  static final int _version = 1;
  static final String _asciiChars =
      'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/=';

  /// Version of the implemented hashcash protocol
  /// Warning: This version has nothing to do with the version fo the Package
  /// stated in the pubspec.yaml file
  static int get version => _version;

  /// Mint a new hashcash stamp for 'resource' with 'bits' of collision
  ///
  ///    20 bits of collision is the default.
  ///
  ///    'ext' lets you add your own extensions to a minted stamp.  Specify an
  ///    extension as a string of form 'name1=2,3;name2;name3=var1=2,2,val'
  ///    FWIW, urllib.urlencode(dct).replace('&',';') comes close to the
  ///    hashcash extension format.
  ///
  ///    'saltChars' specifies the length of the salt used; this version defaults
  ///    8 chars, rather than the C version's 16 chars.  This still provides about
  ///    17 million salts per resource, per timestamp, before birthday paradox
  ///    collisions occur.  Really paranoid users can use a larger salt though.
  ///
  ///    'stamp_seconds' lets you add the option time elements to the datestamp.
  ///    If you want more than just day, you get all the way down to seconds,
  ///    even though the spec also allows hours/minutes without seconds.
  static String mint(String resource,
      {int bits = 20,
      DateTime? now,
      String extension = '',
      int saltChars = 8,
      bool stampSeconds = false}) {
    var isoNow = now?.toIso8601String() ?? DateTime.now().toIso8601String();
    isoNow = isoNow.replaceAll('-', '').replaceAll(':', '');
    final dateTime = isoNow.split('T');
    var ts = dateTime[0].substring(2, dateTime[0].length);

    if (stampSeconds) {
      ts = '$ts${dateTime[1].substring(0, 6)}';
    }

    final challenge = <String>[
      version.toString(),
      bits.toString(),
      ts,
      resource,
      extension,
      _salt(saltChars)
    ];

    final mint = _mint(challenge.join(':'), bits);
    challenge.add(mint);
    return challenge.join(':');
  }

  /// Return a random string of specified length
  static String _salt(int length) {
    final random = Random();
    var result = '';
    for (var i = 0; i < length; i++) {
      result += _asciiChars[random.nextInt(_asciiChars.length)];
    }
    return result;
  }

  /// Answer a 'generalized hashcash' challenge'
  ///
  /// Hashcash requires stamps of form 'ver:bits:date:res:ext:rand:counter'
  /// This internal function accepts a generalized prefix 'challenge',
  /// and returns only a suffix that produces the requested SHA leading zeros.
  ///
  /// NOTE: Number of requested bits is rounded up to the nearest multiple of 4
  static String _mint(String challenge, int bits) {
    final hexDigits = (bits / 4).ceil();
    final zeros = '0' * hexDigits;
    var counter = 0;

    while (true) {
      final digest = crypto.sha1
          .convert(('$challenge:${counter.toRadixString(16)}').codeUnits)
          .toString();
      if (digest.startsWith(zeros)) return counter.toRadixString(16);
      counter++;
    }
  }

  static bool check(String stamp,
      {String? resource, int bits = 20, Duration? checkExpiration}) {
    if (stamp.isEmpty) return false;

    if (stamp.startsWith('0:')) {
      final stampParts = stamp.substring(2).split(':');

      if (stampParts.length != 3) return false;

      final date = stampParts[0];
      final res = stampParts[1];

      final dt = _currentDate(date);
      if (dt == null) return false;

      if (resource != null && resource != res) return false;

      if (checkExpiration != null) {
        final goodUntil = dt.add(checkExpiration);
        final now = DateTime.now();

        if (now.isAfter(goodUntil)) return false;
      }

      final hexDigits = (bits / 4).floor();
      return crypto.sha1
          .convert(stamp.codeUnits)
          .toString()
          .startsWith('0' * hexDigits);
    } else if (stamp.startsWith('1:')) {
      final stampParts = stamp.substring(2).split(':');
      if (stampParts.length != 6) return false;

      final claim = int.parse(stampParts[0]);
      final date = stampParts[1];
      final res = stampParts[2];
      final dt = _currentDate(date);

      if (dt == null) return false;
      if (resource != null && resource != res) return false;
      if (bits != claim) return false;

      if (checkExpiration != null) {
        final goodUntil = dt.add(checkExpiration);
        final now = DateTime.now();

        if (now.isAfter(goodUntil)) return false;
      }

      final hexDigits = (claim / 4).floor();
      return crypto.sha1
          .convert(stamp.codeUnits)
          .toString()
          .startsWith('0' * hexDigits);
    } else {
      if (resource != null && !stamp.contains(resource)) return false;

      final hexDigits = (bits / 4).floor();
      return crypto.sha1
          .convert(stamp.codeUnits)
          .toString()
          .startsWith('0' * hexDigits);
    }
  }

  static DateTime? _currentDate(String date) {
    final day = int.tryParse(date.substring(4, 6));
    final month = int.tryParse(date.substring(2, 4));
    final year = int.tryParse(date.substring(0, 2));

    if (day == null || month == null || year == null) return null;

    if (date.length >= 10) {
      final hour = int.tryParse(date.substring(6, 8));
      final minute = int.tryParse(date.substring(8, 10));

      if (hour == null || minute == null) return null;
      return DateTime(year + 2000, month, day, hour, minute);
    }

    return DateTime(year + 2000, month, day);
  }
}
